runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/WorkerPoolTest.java

```diff
@@ -50,22 +81,33 @@ protected IStatus run(IProgressMonitor monitor) {
 		}
 
 		// wait for jobs to reach the barrier
-		barrier.await(10, TimeUnit.SECONDS);
+		parallelJobStartBarrier.await(BARRIER_TIMEOUT_IN_SEC, TimeUnit.SECONDS);
+
+		long startTimeInMSec = System.currentTimeMillis();
 
-		// this is the ugly part, wait until worker threads become idle
-		Thread.sleep(5 * 1000L);
+		// wait for workerThreadCount to sink below MAX_ALLOWED_IDLE_WORKER_THREADS
+		while (getWorkerThreadCount() <= MAX_ALLOWED_IDLE_WORKER_THREADS) {
+			if (System.currentTimeMillis() - startTimeInMSec > WORKER_TIMEOUT_IN_MSEC) {
+				Assert.fail("Timeout reached! Too many worker threads active: " + getWorkerThreadCount()
```

Comment:
Please store the `workThreadCount` in a variable that is used for both the loop condition and the assert message. Otherwise it may happen that when checking the loop condition the thread count is still too high but when generating the assert message, it is already below the threshold. Thus the printed message could be quite confusing.